persist/fs/seek_manager.go

```diff
@@ -137,75 +170,198 @@ func (m *seekerManager) CacheShardIndices(shards []uint32) error {
+	byTime.seekers[startNano] = seekersAndBloom
+	byTime.Unlock()
+	return availableSeeker.seeker, nil
 }
 
-func (m *seekerManager) Seeker(shard uint32, start time.Time) (FileSetSeeker, error) {
+func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentFileSetSeeker) error {
 	byTime := m.seekersByTime(shard)
 
 	byTime.Lock()
-	// Track accessed to precache in open/close loop
-	byTime.accessed = true
 
 	startNano := xtime.ToUnixNano(start)
-	seeker, ok := byTime.seekers[startNano]
-	if ok {
+	seekersAndBloom, ok := byTime.seekers[startNano]
+	// Should never happen - This either means that the caller (BlockRetriever) is trying to return seekers
+	// that it never requested, OR its trying to return seekers after the openCloseLoop has already
+	// determined that they were all no longer in use and safe to close. Either way it indicates there is
+	// a bug in the code.
+	if !ok {
 		byTime.Unlock()
-		return seeker, nil
+		return errSeekersDontExist
 	}
 
-	var err error
-	seeker, err = m.newOpenSeeker(shard, start)
-	if err == nil {
-		byTime.seekers[startNano] = seeker
+	found := false
+	for i, compareSeeker := range seekersAndBloom.seekers {
+		if seeker == compareSeeker.seeker {
+			found = true
+			compareSeeker.isBorrowed = false
+			seekersAndBloom.seekers[i] = compareSeeker
+			break
+		}
 	}
+	// Should never happen with a well behaved caller. Either they are trying to return a seeker
+	// that we're not managing, or they provided the wrong shard/start.
+	if !found {
+		return errReturnedUnmanagedSeeker
+	}
+
+	byTime.seekers[startNano] = seekersAndBloom
 	byTime.Unlock()
+	return nil
+}
+
+// getOrOpenSeekersWithLock checks if the seekers are already open / initialized. If they are, then it
+// returns them. Then, it checks if a different goroutine is in the process of opening them , if so it
+// registers itself as waiting until the other goroutine completes. If neither of those conditions occur,
+// then it begins the process of opening the seekers itself. First, it creates a waitgroup that other
+// goroutines can use so that they're notified when the seekers are open. This is useful because it allows
+// us to prevent multiple goroutines from trying to open the same seeker without having to hold onto a lock
+// of the seekersByTime struct during a I/O heavy workload. Once the wg is created, we relinquish the lock,
+// open the Seeker (I/O heavy), re-acquire the lock (so that the waiting goroutines don't get it before us),
+// and then notify the waiting goroutines that we've finished.
+func (m *seekerManager) getOrOpenSeekersWithLock(start xtime.UnixNano, byTime *seekersByTime) (seekersAndBloom, error) {
+	seekers, ok := byTime.seekers[start]
+	if ok && seekers.wg == nil {
+		// Seekers are already open
+		return seekers, nil
+	}
+
+	if seekers.wg != nil {
+		// Seekers are being initialized / opened, wait for the that to complete
+		byTime.Unlock()
+		seekers.wg.Wait()
```

Comment:
Seems racy to perform `seekers.wg.Wait()` outside of lock as it may be nil'd out between the Unlock and the wait call no?

Reply:
Hm, actually seekers is just a value struct not a ref here?  Maybe it's safe in that case.

Reply:
yeah you had me worried for a second, but because its a value struct it should be fine. When the .wg field is set to nil, our's will still point to the wg and when we call .Wait() on it we should just proceed immediately